src/sentry/auth/email.py

```diff
@@ -10,15 +13,97 @@ def __init__(self, email: str, users: Sequence[User]) -> None:
+            raise ValueError
+        if len(candidates) == 1:
+            (unique_email,) = candidates
+            return unique_email.user
+
+        for step_cls in self.get_steps():
+            step = step_cls(self)
+            last_candidates = candidates
+            candidates = tuple(step.apply(candidates))
+            if len(candidates) == 1:
+                # Success: We've narrowed down to only one candidate
+                (choice,) = candidates
+                step.if_conclusive(last_candidates, choice)
+                return choice.user
+            if len(candidates) == 0:
+                # If the step eliminated all, ignore it and go to the next step
+                candidates = last_candidates
+
+        self.if_inconclusive(candidates)
+        raise AmbiguousUserFromEmail(self.email, [ue.user for ue in candidates])
+
+    def if_inconclusive(self, remaining_candidates: Sequence[UserEmail]) -> None:
+        """Hook to call if no step resolves to a single user."""
+        metrics.incr("auth.email_resolution.no_resolution", sample_rate=1.0)
+
+    @dataclass
+    class ResolutionStep(abc.ABC):
+        parent: "_EmailResolver"
+
+        @abc.abstractmethod
+        def apply(self, candidates: Sequence[UserEmail]) -> Iterable[UserEmail]:
+            raise NotImplementedError
+
+        def if_conclusive(self, candidates: Sequence[UserEmail], choice: UserEmail) -> None:
+            """Hook to call if this step resolves to a single user."""
+            pass
+
+    class IsVerified(ResolutionStep):
+        """Prefer verified email addresses."""
+
+        def apply(self, candidates: Sequence[UserEmail]) -> Iterable[UserEmail]:
+            return (ue for ue in candidates if ue.is_verified)
+
+        def if_conclusive(self, candidates: Sequence[UserEmail], choice: UserEmail) -> None:
+            metrics.incr("auth.email_resolution.by_verification", sample_rate=1.0)
+
+    class HasOrgMembership(ResolutionStep):
+        """Prefer users who belong to the organization."""
+
+        def apply(self, candidates: Sequence[UserEmail]) -> Iterable[UserEmail]:
+            if not self.parent.organization:
+                return ()
+            query = OrganizationMember.objects.filter(
+                organization=self.parent.organization, user__in=[ue.user for ue in candidates]
+            )
+            users_in_org = {user_id for (user_id,) in query.values_list("user")}
+            return (ue for ue in candidates if ue.user.id in users_in_org)
+
+        def if_conclusive(self, candidates: Sequence[UserEmail], choice: UserEmail) -> None:
+            metrics.incr("auth.email_resolution.by_org_membership", sample_rate=1.0)
+
+    class IsPrimary(ResolutionStep):
+        """Prefer users whose primary address matches the address in question."""
+
+        def apply(self, candidates: Sequence[UserEmail]) -> Iterable[UserEmail]:
+            return (ue for ue in candidates if ue.user.email == self.parent.email)
```

Comment:
i'm not sure if this logic is correct. we are checking the UserEmail user's email against the parent email, but not that the UserEmail's email matches the UserEmail user's email. (say that 3 times fast!!). I think we need to check that the `UserEmail.is_primary() == True` as well here.

I think that this would not resolve correctly currently if there were a user with multiple secondary emails and we were attempting to resolve the primary.

I also wonder if UserEmail is case normalized

Reply:
I want to say the logic checks out because `self.parent.email` is already guaranteed to be the UserEmail's email because that's what we queried for; that is, it's equivalent to writing `if ue.user.email == ue.email`. But you're right, it's much better to just call `if ue.is_primary()` (which -- hey! -- is also just `return self.user.email == self.email`).

As for being case-normalized, yeah, that looks messy. We seem to habitually write queries like `email__iexact` to treat it as case-insensitive, but `is_primary` doesn't respect that. It probably needs a fix, but one that would go on the `UserEmail` model class.